insalan/mailer.py

```diff
@@ -0,0 +1,255 @@
+
+    def send_kick_mail(self, user_object: User, team_name: str):
+        """
+        Send a mail to a user that has been kicked.
+        """
+        connection = get_connection(
+            fail_silently=False,
+            username=self.MAIL_FROM,
+            password=self.MAIL_PASS,
+            host=self.MAIL_HOST,
+            port=self.MAIL_PORT,
+            use_ssl=self.MAIL_SSL,
+        )
+        email = EmailMessage(
+            insalan.settings.EMAIL_SUBJECT_PREFIX + _("Vous avez été exclu.e de votre équipe"),
+            _("Vous avez été exclu.e de l'équipe %s.") % team_name,
+            self.MAIL_FROM,
+            [user_object.email],
+            connection=connection,
+        )
+        if self.TEST:
+            email.send()
+        else:
+            self.queue.append(email)
+
+    def send_ticket_mail(self, user_object: User, ticket: str):
+        """
+        Send a mail to a user that has been kicked.
+        """
+
+        ticket_pdf = TicketManager.generate_ticket_pdf(ticket)
+
+        connection = get_connection(
+            fail_silently=False,
+            username=self.MAIL_FROM,
+            password=self.MAIL_PASS,
+            host=self.MAIL_HOST,
+            port=self.MAIL_PORT,
+            use_ssl=self.MAIL_SSL,
+        )
+        email = EmailMessage(
+            insalan.settings.EMAIL_SUBJECT_PREFIX + _("Votre billet pour l'InsaLan"),
+            _("Votre inscription pour l'Insalan a été payée. Votre billet est disponible en pièce jointe. Vous pouvez retrouver davantages d'informations sur l'évènement sur le site internet de l'InsaLan."),
+            self.MAIL_FROM,
+            [user_object.email],
+            connection=connection,
+        )
+        email.attach(
+            TicketManager.create_pdf_name(ticket),
+            ticket_pdf,
+            "application/pdf"
+        )
+
+        if self.TEST:
+            email.send()
+        else:
+            self.queue.append(email)
+
+    def send_tournament_mail(self, user_object: User, title: str, content: str, attachment: str):
+        """
+        Send a mail
+        """
+        connection = get_connection(
+            fail_silently=False,
+            username=self.MAIL_FROM,
+            password=self.MAIL_PASS,
+            host=self.MAIL_HOST,
+            port=self.MAIL_PORT,
+            use_ssl=self.MAIL_SSL,
+        )
+        email = EmailMessage(
+            insalan.settings.EMAIL_SUBJECT_PREFIX + title,
+            content,
+            self.MAIL_FROM,
+            [user_object.email],
+            connection=connection,
+        )
+        if attachment:
+            email.attach(attachment.name, attachment.read())
+
+        if self.TEST:
+            email.send()
+        else:
+            self.queue.append(email)
+
+    def send_first_mail(self):
+        """
+        Send the first mail in the queue.
+        """
+        if len(self.queue) == 0:
+            return
+        self.queue[0].send()
```

Comment:
if there any control on whether or not the sending failed? It's probably asynchronous, but I thought the purpose of the mailer was to be able to send emails again/retry

Reply:
The purpose of the mailer is actually to :
- Handle multiple email addresses (contact and tournament for now).
- Don't reach our mail quotas when sending tournaments mails to every players.

We can maybe add a mail back to the queue when it fail but that require to have a max number of retry to not overwhelm the mailer with mails that we can't send.

what's your thoughts on this ?